include/mockturtle/generators/arithmetic.hpp

```diff
@@ -183,4 +226,85 @@ inline std::vector<signal<Ntk>> carry_ripple_multiplier( Ntk& ntk, std::vector<s
   return res;
 }
 
+/*! \brief Creates a sideways sum adder using half and full adders
+ *
+ * The function creates the adder in `ntk` and returns output signals,
+ * whose size is floor(log2(a.size()))+1.
+ *
+ * \param ntk Network
+ * \param a Input operand
+ */
+template<typename Ntk>
+inline std::vector<signal<Ntk>> sideways_sum_adder( Ntk& ntk, std::vector<signal<Ntk>> const& a )
+{
+  static_assert( is_network_type_v<Ntk>, "Ntk is not a network type" );
+
+  int n = a.size();
+
+  int out_n = 1; // floor(log2(n) + 1)
+  int tmpn = n;
+  while( tmpn >>= 1 ) out_n++;
+
+  std::list<signal<Ntk>> *first_level  = new std::list<signal<Ntk>>; // sum bits
```

Comment:
Is there a reason you have `first_level` and `second_level` on the heap? You could just write `std::list<signal<Ntk>> first_level;`. Then there is no need to delete it in the end.

Reply:
The reason is because I am swapping them in lines 298-300, but I can do that in some other way.

Reply:
Understand. Didn't spot that before. But I think that https://en.cppreference.com/w/cpp/container/list/swap efficiently implements swapping two lists.